### PR TITLE
feat: implement missing encodings and refactor

### LINKED ISSUE #15: A complete implementation of encodings in the multibase table
Please see [implementation](https://github.com/PolkaX/rust-ipfs/tree/develop/core/multibase) for details.

@dignifiedquire If you think there is no problem with the implementation, I will open a PR.

### PR DESCRIPTION
**Pull request by koushiro**
Signed-off-by: koushiro <koushiro.cqx@gmail.com>

* Implement all multibase encodings
* Refactor lib and cli

Close #15

### CODE REVIEW

**Comment by koushiro** [comment]
#Benchmark:

## Before

```
base_x                  time:   [350.14 us 358.97 us 368.72 us]
base32                  time:   [2.9558 us 2.9993 us 3.0583 us]
Found 5 outliers among 100 measurements (5.00%)
  4 (4.00%) high mild
  1 (1.00%) high severe
base64                  time:   [798.46 ns 803.70 ns 809.57 ns]
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe

base_x #2               time:   [64.800 us 64.992 us 65.201 us]
base32 #2               time:   [2.9455 us 2.9594 us 2.9757 us]
Found 7 outliers among 100 measurements (7.00%)
  3 (3.00%) high mild
  4 (4.00%) high severe
base64 #2               time:   [807.48 ns 809.73 ns 812.26 ns]
Found 8 outliers among 100 measurements (8.00%)
  8 (8.00%) high mild
```

## After: 

```
encode/base32           time:   [904.17 ns 905.82 ns 908.09 ns]
Found 4 outliers among 100 measurements (4.00%)
  2 (2.00%) high mild
  2 (2.00%) high severe
encode/base58btc        time:   [331.32 us 331.44 us 331.59 us]
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) high mild
  2 (2.00%) high severe
encode/base64           time:   [739.74 ns 740.53 ns 741.41 ns]
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe

decode/base32           time:   [817.73 ns 818.51 ns 819.38 ns]
Found 6 outliers among 100 measurements (6.00%)
  4 (4.00%) high mild
  2 (2.00%) high severe
decode/base58btc        time:   [67.511 us 67.550 us 67.589 us]
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe
decode/base64           time:   [775.68 ns 776.13 ns 776.62 ns]
Found 9 outliers among 100 measurements (9.00%)
  1 (1.00%) low mild
  2 (2.00%) high mild
  6 (6.00%) high severe
```

File: cli/main.rs [hunk 1]
```diff
@@ -17,65 +14,87 @@ enum Mode {
...
+        Mode::Decode { input } => decode(&input),
     }
 }
 
 #[derive(Debug)]
 struct StrBase(Base);
 
-impl fmt::Display for StrBase {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+impl std::fmt::Display for StrBase {
```
Comment (vmx) [inline cli/main.rs]: Please try to keep changes that are not related to the actual work to a minimum. In this case, please keep the original imports.

File: cli/main.rs [hunk 2]
```diff
@@ -87,26 +106,16 @@ impl From<StrBase> for Base {
...
-    log::debug!("encoding with {}", base);
-    let mut stdin = io::stdin();
-    let mut stdout = io::stdout();
-    let mut buffer = Vec::new();
-    stdin.read_to_end(&mut buffer)?;
-    log::debug!("read {:?} from stdin", buffer);
-    let result = multibase::encode(base.into(), buffer.as_slice());
-    stdout.write_all(result.as_bytes())?;
+fn encode(base: StrBase, input: &[u8]) -> Result<()> {
+    println!("Encode {:?} with {}", input, base);
```
Comment (vmx) [inline cli/main.rs]: Please use the logging functionality like `log::debug!()` instead of using `println!()`

File: src/encoding.rs [hunk 3]
```diff
@@ -0,0 +1,87 @@
+use data_encoding::{Encoding, Specification};
+use lazy_static::lazy_static;
+
+fn make_encoding(symbols: &str, padding: Option<char>) -> Encoding {
```
Comment (vmx) [inline src/encoding.rs]: Recently https://github.com/multiformats/rust-multibase/pull/16 was merged.

There we use https://docs.rs/data-encoding-macro/0.1.7/data_encoding_macro/ for custom encodings.

See https://github.com/multiformats/rust-multibase/blob/4220b42bd422b940d46fd05589c2672e861f63a1/src/base.rs#L100-L103 for an example.

I think it would make sense to use it here as well.

File: src/error.rs [hunk 4]
```diff
@@ -1,40 +1,34 @@
...
-}
-
-impl error::Error for Error {
-    fn description(&self) -> &str {
-        use Error::*;
-
-        match *self {
-            UnknownBase => "Unknown base",
-            InvalidBaseString => "Invalid base string",
+impl std::fmt::Display for Error {
```
Comment (vmx) [inline src/error.rs]: Please keep the original import `use std::{error, fmt}` instead of using the fully qualified name.

File: src/error.rs [hunk 5]
```diff
@@ -1,40 +1,34 @@
...
         }
     }
 }
 
+impl std::error::Error for Error {}
+
 impl From<base_x::DecodeError> for Error {
     fn from(_: base_x::DecodeError) -> Self {
-        Self::InvalidBaseString
+        Error::InvalidBaseString
```
Comment (vmx) [inline src/error.rs]: Please keep the current style of using `Self` instead of `Error`.

File: tests/tests.rs [hunk 6]
```diff
@@ -0,0 +1,107 @@
+use multibase::{decode, encode, Base, Base::*};
```
Comment (vmx) [inline tests/tests.rs]: Could you please rename the file back to `tests/lib.rs`. I don't see a need to change that.

File: Cargo.toml [hunk 7]
```diff
@@ -1,21 +1,21 @@
 [package]
 name = "multibase"
 version = "0.7.0"
-description = "multibase rust"
-homepage = "https://github.com/multiformats/rust-multibase"
-authors = ["Friedel Ziegelmayer <dignifiedquire@gmail.com>"]
+authors = [
+    "Friedel Ziegelmayer <dignifiedquire@gmail.com>",
+    "Qinxuan Chen <koushiro.cqx@gmail.com>",
```
Comment (vmx) [inline Cargo.toml]: I'd like to reserve the "authors" for people that have also permission to publish a crate. This should be an entry point for people that want to get in touch with the maintainers.

File: src/encoding.rs (hunk 3 above)
Comment (koushiro) [reply src/encoding.rs]: I don't want to use the `data_encoding_macro`, which introduces so much dependencies.

Comment (vmx) [reply src/encoding.rs]: Though you wouldn't need the `lazy_static` dependency. The dependencies `data_encoding_macro` introduces are well established ones though, that you'd need as soon as you use macros.

File: src/lib.rs [hunk 8]
```diff
@@ -1,44 +1,79 @@
...
+///
+/// ```
+/// use multibase::decode_base58btc;
+///
+/// assert_eq!(
+///     decode_base58btc("Cn8eVZg").unwrap(),
+///     b"hello".to_vec(),
+/// );
+/// ```
+pub fn decode_base58btc<I: AsRef<str>>(input: I) -> Result<Vec<u8>> {
```
Comment (vmx) [inline src/lib.rs]: Could you please remove `decode_base58btc()` and `encode_base58btc()`? I don't want to introduce a broader public API for one special case.

File: src/encoding.rs (hunk 3 above)
Comment (koushiro) [reply src/encoding.rs]: The dependency `data_encoding_macro` uses syn 0.15 and I don't want to introduce it into my dependency graph. I will open a PR for `data-encoding`, and once it is merged, I will consider using the `data_encoding_marco`

File: tests/lib.rs [hunk 9]
```diff
@@ -1,149 +1,107 @@
...
-        (Base58btc, id2.to_vec())
-    );
-
-    assert_eq!(decode("mZg").unwrap(), (Base64UpperNoPad, b"f".to_vec()));
-    assert_eq!(decode("MZg==").unwrap(), (Base64UpperPad, b"f".to_vec()));
-    assert_eq!(decode("uZg").unwrap(), (Base64UrlUpperNoPad, b"f".to_vec()));
-    assert_eq!(decode("UZg==").unwrap(), (Base64UrlUpperPad, b"f".to_vec()));
+fn preserves_leading_zeroes() {
+    let input = b"\x00\x00\x00yes mani !";
+    println!("{:x?}", input.to_vec());
```
Comment (vmx) [inline tests/lib.rs]: Nitpick: this looks like a debug left-over.

(Nitpicks are things that I found, but are not worth changing if it is the only thing to change or if out feel like it's not worth the effort)

File: src/encoding.rs (hunk 3 above)
Comment (vmx) [reply src/encoding.rs]: Not introducing a large dependency tree is surely reasonable. I'm OK with taking your current version. We can introduce `data_encoding_macro` later on if we like.

Comment (koushiro) [reply src/encoding.rs]: Thanks for your review, if `data_encoding_macro` releases a new version, I will use it.

File: src/lib.rs (hunk 8 above)
Comment (koushiro) [reply src/lib.rs]: At the current stage, I think the `multibase` was developed for [rust-cid](https://github.com/multiformats/rust-cid) only, you can see [this](https://github.com/multiformats/rust-cid/blob/master/src/to_cid.rs#L50) for detail about why I add `decode_base58btc` and `encode_base58btc`.

File: src/encoding.rs (hunk 3 above)
Comment (vmx) [reply src/encoding.rs]: @koushiro: it seems that `data_encoding_macro` was already updated to syn 1.0. But there wasn't a release yet.

Comment (koushiro) [reply src/encoding.rs]: Yeah, I found it :(

File: src/lib.rs (hunk 8 above)
Comment (vmx) [reply src/lib.rs]: Good thinking! What if we make the decoders part of the public API, so that you can call `Base58Btc::decode()` directly?

Comment (koushiro) [reply src/lib.rs]: I found that we can use `Base::Base58Btc.encode()` and `Base::Base58Btc.decode()` directly for CIDv0 specially, so I removed `decode_base58btc`  and `encode_base58btc`.

File: Cargo.toml [hunk 10]
```diff
@@ -1,21 +1,18 @@
...
+description = "multibase in rust"
+homepage = "https://github.com/multiformats/rust-multibase"
+keywords = ["ipld", "ipfs", "multihash", "multibase", "cid"]
 
 [dependencies]
 base-x = "0.2"
-data-encoding = "2.1.2"
-data-encoding-macro = "0.1.7"
+data-encoding = "2.1"
+lazy_static = "1.4"
```
Comment (cyphar) [inline Cargo.toml]: Is there a reason to not do

```suggestion
data-encoding = "^2.1"
lazy_static = "^1.4"
```

Comment (cyphar) [inline Cargo.toml]: Is there a reason to not do

```suggestion
base-x = "^0.2"
data-encoding = "^2.1"
lazy_static = "^1.4"
```

File: cli/Cargo.toml [hunk 11]
```diff
@@ -1,17 +1,18 @@
...
-env_logger = "0.6.2"
-exitfailure = "0.5.1"
-failure = "0.1.5"
-log = "0.4.8"
+anyhow = "1.0"
+env_logger = "0.7"
+log = "0.4"
 multibase = { path = ".." }
-structopt = "0.3.0"
+structopt = "0.3"
```
Comment (cyphar) [inline cli/Cargo.toml]: Same question here -- why not do `^` for all versions?

File: Cargo.toml (hunk 10 above)
Comment (koushiro) [reply Cargo.toml]: Because it's the same

Comment (cyphar) [reply Cargo.toml]: `^1.4` isn't the same as `1.4` -- `^1.4` will allow automatically updating to use `1.9999999`, while `1.4` won't.

Comment (koushiro) [reply Cargo.toml]: See [doc](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#specifying-dependencies-from-cratesio) for details.

> The string "0.1.12" is a semver version requirement. Since this string does not have any operators in it, it is interpreted the same way as if we had specified "^0.1.12", which is called a caret requirement.

File: cli/Cargo.toml (hunk 11 above)
Comment (koushiro) [reply cli/Cargo.toml]: ditto

File: Cargo.toml (hunk 10 above)
Comment (cyphar) [reply Cargo.toml]: That's true for pre-`1.0` projects, but isn't true for post-1.0 projects. From the same link:

> If instead we had specified the version string as ^1.0, cargo should update to 1.1 if it is the latest 1.y release, but not 2.0.

Comment (koushiro) [reply Cargo.toml]: "0.1.12" is just a example...
What the document actually wants to express is that `"x.y.z"` is interpreted  the same way as if we had specified `"^x.y.z"`

Comment (cyphar) [reply Cargo.toml]: I quoted the document you linked where it explicitly explains how caret requirements work. It *does not* say that caret requirements are a no-op -- what use would the feature have if it did nothing? If you don't believe me, try build a crate that has dependencies like `lazy_static = "^1.0"` -- it will use `lazy_static 1.4.0`:

```
% grep lazy_static Cargo.toml
lazy_static = "^1.0"
% cargo build
    Updating crates.io index
   Compiling lazy_static v1.4.0
   Compiling test1 v0.1.0 (/home/cyphar/tmp/test1)
    Finished dev [unoptimized + debuginfo] target(s) in 4.86s
```

But okay, I was only making a suggestion. If you don't want to do it, that's fine.

Comment (koushiro) [reply Cargo.toml]: Have you tried to build a crate with lazy_static=1.0? It will use lazy_static v1.4 too.

Comment (cyphar) [reply Cargo.toml]: Oh... You're quite right. Oops.

File: src/encoding.rs (hunk 3 above)
Comment (koushiro) [reply src/encoding.rs]: The `data-encoding-macro` v0.1.8 has been released, so I used it instead of `lazy_static`.
